walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requestAttendees = request.getAttendees();
+    List<TimeRange> res = new LinkedList<>();
+    res.add(TimeRange.WHOLE_DAY);
+    for (Event e : events) {
+      Set<String> Attendees = e.getAttendees();
+      boolean flag = false;
+      for (String a : Attendees) {
+        if (requestAttendees.contains(a)) {
+          flag = true;
+          break;
+        }
+      }
+      if (!flag){
+        continue;
+      }
+      TimeRange badTime = e.getWhen();
+      ListIterator it = res.listIterator();
+      while (it.hasNext()) {
+        TimeRange goodTime = (TimeRange) it.next();
+        if (goodTime.start() >= badTime.start() + badTime.duration())
+          continue;
+        if (goodTime.start() < badTime.start()) {
+          if (badTime.start() >= goodTime.start() + goodTime.duration()) {
+            continue;
```

Comment:
You can use the TimeRange.overlaps function here:

https://github.com/nonstopfor/my-portfolio/blob/week4/walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/TimeRange.java#L82